File: src/multiplex.c
/* The MIT License (MIT)
 * 
 * Copyright (c) 2013 Yannick Scherer
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <fcntl.h>
#include "multiplex.h"

// -- Buffer Structure
typedef struct ChannelBuffer {
    char * data;    // receive buffer
    int offset;     // current read offset
    int length;     // current read length
    int capacity;   // capacity of receive buffer
    char newData;   // 0 = no new data since last 'select'
} ChannelBuffer;

// -- CREATE
Multiplex * multiplex_new(int fd) {
    Multiplex * m = (Multiplex *)calloc(1, sizeof(Multiplex));
    if (m != 0) {
        m->fd = fd;
    }
    return m;
}

// -- ACTIVATE CHANNEL
void multiplex_activate_channel(Multiplex * c, unsigned char channelId, int initialBufferSize) {
    if (c == 0 || c->channels[channelId] != 0) return;
    else {
        ChannelBuffer * buf = (ChannelBuffer *)calloc(1, sizeof(ChannelBuffer));
        if (buf != 0) {
            int size = initialBufferSize > 0 ? initialBufferSize : CHANNEL_INITIAL_BUFFER_SIZE;
            char * stream = (char *)calloc(size, sizeof(char));
            if (stream == 0) return;
            buf->data = stream;
            buf->offset = 0;
            buf->length = 0;
            buf->capacity = size;
            c->channels[channelId] = buf;
        }
    }
}

// -- REALLOCATE BUFFER
static int multiplex_reallocate_channel(Multiplex * c, unsigned char channelId, int additionalDataSize) {
    if (c == 0 || c->channels[channelId] == 0) return 0;
    else {
        ChannelBuffer * buf = c->channels[channelId];
        char * tmpBuf = buf->data;
        int newLen = buf->offset + buf->length + additionalDataSize;
        int allocateLen = buf->capacity;

        // Case 1: buffer is big enough
        if (allocateLen >= newLen) return 1;

        // Case 2: move data within buffer (set offset to 0)
        if (buf->capacity >= buf->length + additionalDataSize) {
            if (buf->offset > 0) {
                int i = 0; 
                while (i < buf->length) {
                    buf->data[i] = buf->data[buf->offset + i];
                }
                buf->offset = 0;
            }
            return 1;
        }

        // Case 3: extend buffer
        while (allocateLen < newLen) allocateLen *= 2;
        buf->data = (char *)calloc(allocateLen, sizeof(char));
        if (buf->data == 0) { buf->data = tmpBuf; return 0; }
        buf->capacity = allocateLen;
        memcpy(buf->data, tmpBuf + buf->offset, buf->length);
        free(tmpBuf);
        return 1;
    }
}

// -- WRITE TO/READ FROM BUFFER
void multiplex_write(Multiplex * c, unsigned char channelId, char * data, int offset, int length) {
    if (!multiplex_reallocate_channel(c, channelId, length)) return;
    else {
        ChannelBuffer * buf = c->channels[channelId];
        if (buf == 0) return;
        memcpy(buf->data + buf->offset + buf->length, data + offset, length);
        buf->length += length;
        buf->newData = length;
    }
}

int multiplex_read(Multiplex * c, int channelId, char * dst, int offset, int length) {
    if (c == 0 || c->channels[channelId] == 0) return CHANNEL_CLOSED;
    else {
        ChannelBuffer * buf = c->channels[channelId];
        int copyLen = length;
        if (buf == 0) return CHANNEL_IGNORED;
        if (buf->length < length) copyLen = buf->length;
        memcpy(dst + offset, buf->data + buf->offset, copyLen);
        buf->offset += copyLen;
        buf->length -= copyLen;
        buf->newData -= copyLen;
        if (buf->newData < 0) buf->newData = 0;
        return copyLen;
    }
}

// -- READ FIXED LENGTH
static int _fd_read(int fd, int timeoutMs, char * buffer, int length) {
    int position = 0, bytesRead = 0, selectResult = -1;
    struct timeval timeout;
    fd_set fds;

    while (position < length) {
        FD_ZERO(&fds);
        FD_SET(fd, &fds);
        timeout.tv_sec = timeoutMs / 1000;
        timeout.tv_usec = (timeoutMs % 1000) * 1000;

        selectResult = select(fd + 1, &fds, (fd_set *)0, (fd_set *)0, &timeout);
        if (selectResult == 0) return CHANNEL_TIMEOUT;

        if (FD_ISSET(fd,&fds)) {
            bytesRead = read(fd, buffer + position, length - position);
            if (bytesRead <= 0) return CHANNEL_CLOSED;
            else position += bytesRead;
        }
    }
    return length;
}

// -- READ FROM PIPE (returns channel ID as int32; -1 for ignored data, -255 for closed pipe)
int multiplex_select(Multiplex * c, int timeoutMs) {
    char prefixBuffer[5];
    int bytesRead = 0, dataLength = 0;
    unsigned char channelId = 0;

    //
    if (c == 0) return CHANNEL_CLOSED;

    // Check if data is available somewhere
    {
        int i = 0;
        ChannelBuffer * buf = 0;
        for (; i < 256; ++i) {
            buf = c->channels[i];
            if (buf != 0 && buf->length > 0 && buf->newData != 0) {
                buf->newData = 0;
                return i;
            }
        }
    }

    //
    bytesRead = _fd_read(c->fd, timeoutMs, prefixBuffer, 5); 
    if (bytesRead != 5) return bytesRead;

    //
    dataLength = (prefixBuffer[0] << 24) | (prefixBuffer[1] << 16) | (prefixBuffer[2] << 8) | prefixBuffer[3];
    channelId = (unsigned char)prefixBuffer[4];
    {
        char buffer[dataLength - 1];
        bytesRead = _fd_read(c->fd, timeoutMs, buffer, dataLength - 1);
        if (bytesRead != dataLength - 1) return bytesRead;
        if (c->channels[channelId] == 0) return CHANNEL_IGNORED;
        multiplex_write(c, channelId, buffer, 0, dataLength - 1);
    }
    return channelId;
}

// -- RECEIVE
int multiplex_receive(Multiplex * c,
                      int timeoutMs,
                      int channelId,
                      char * dst,
                      int length) {
    int receiveId = CHANNEL_IGNORED;
    ChannelBuffer * buf = c == 0 ? 0 : c->channels[channelId];
    if (c == 0) return CHANNEL_CLOSED;

    // Check if data is already buffered.
    if (buf != 0 && buf->length > 0) {
        if (length <= buf->length) {
            memcpy(dst, buf->data + buf->offset, length);
            buf->offset += length;
            buf->length -= length;
            return length;
        } else {
            int tmpLen = buf->length;
            memcpy(dst, buf->data + buf->offset, tmpLen);
            buf->offset = 0;
            buf->length = 0;
            return tmpLen;
        }
    }

    // Receive on the given Channel
    receiveId = multiplex_select(c, timeoutMs);
    if (receiveId < 0) return receiveId;
    if (receiveId != channelId) return CHANNEL_IGNORED;

    // Copy from ChannelBuffer
    return multiplex_read(c, channelId, dst, 0, length);
}

// -- SEND
int multiplex_send(Multiplex * c, int channelId, char const * src, int length) {
    const int len = length + 1;
    char buffer[5 + length];
    buffer[0] = (char)((len >> 24) & 0xFF);
    buffer[1] = (char)((len >> 16) & 0xFF);
    buffer[2] = (char)((len >> 8) & 0xFF);
    buffer[3] = (char)(len & 0xFF);
    buffer[4] = (char)(channelId & 0xFF);
    memcpy(buffer + 5, src, length);
    return write(c->fd, buffer, 5 + length);
}

// -- ACCESS
int multiplex_length(Multiplex * c, int channelId) {
    if (c == 0 || c->channels[channelId] == 0) return -1;
    return c->channels[channelId]->length;
}

int multiplex_last_received(Multiplex * c, int channelId) {
    if (c == 0 || c->channels[channelId] == 0) return 0;
    return c->channels[channelId]->newData;
}

char const * multiplex_get(Multiplex * c, int channelId) {
    if (c == 0 || c->channels[channelId] == 0) return 0;
    else {
        ChannelBuffer * buf = c->channels[channelId];
        return buf->data + buf->offset;
    }
}

char * multiplex_copy(Multiplex * c, int channelId, int offset, int length) {
    if (c == 0 || c->channels[channelId] == 0) return 0;
    else {
        ChannelBuffer * buf = c->channels[channelId];
        int dataLen = buf->length - offset;
        int copyLen = dataLen <= length ? dataLen : length;
        char * tmp = (char *)calloc(length, sizeof(char));
        if (tmp == 0) return 0;
        memcpy(tmp, buf->data + buf->offset + offset, copyLen);
        return tmp;
    }
}

void multiplex_clear(Multiplex * c, int channelId) {
    if (c == 0 || c->channels[channelId] == 0) return;
    else {
        ChannelBuffer * buf = c->channels[channelId];
        buf->offset = 0;
        buf->length = 0;
        buf->newData = 0;
    }
}

File: go/multiplex.go
package multiplex

/* The MIT License (MIT)
 *
 * Copyright (c) 2013 Yannick Scherer
 * Copyright (c) 2014 Raffaele Sena
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

// ----------------------------------------------------------------------
//
//   CONCEPT
//
// ----------------------------------------------------------------------
// We assume that packet fragments arrive in-order on the given file
// descriptor. By prefixing a packet with 4 bytes of length (aligned
// right, zero left-padded, includes length of channel ID) and a single
// byte containing the channel ID, we can reassemble the packet and
// decide which channel it belongs to.
//
// Channels have to be activated before use, creating a receive buffer
// that is dynamically extended and reduced when needed. Using a
// blocking 'select' function (with a timeout), we can retrieve the
// ID of a channel that has new data available.
//
// Error and status codes (e.g. "data was received on a channel that
// is not active") are negative, while channel IDs are positive or zero.

import (
	"io"
	"log"
	"net"
	"sync"
	"time"
)

const (
	INITIAL_BUFFER_SIZE = 256
	MAX_CHANNELS        = 256

	headerLength = 5 // 6 // 1:magic + 4:size + 1:channel
	magic        = 0x69
)

type MultiplexError string

func (e MultiplexError) Error() string {
	return "multiplex error: " + string(e)
}

var (
	CHANNEL_IGNORED = MultiplexError("channel ignored")
	CHANNEL_TIMEOUT = MultiplexError("channel timeout")
	CHANNEL_CLOSED  = MultiplexError("channel closed")
)

type ChannelBuffer struct {
	data    []byte // receive buffer
	offset  int    // current read offset
	length  int    // current read length
	initial int    // minimum capacity
	newData int    // 0 = no new data since last 'select'
}

type Multiplex struct {
	conn         net.Conn                     // network connection
	max_channels uint                         // maximum number of channels (0 <= max_channels <= MAX_CHANNELS)
	channels     [MAX_CHANNELS]*ChannelBuffer // O(1) lookup for channels

	sync.Mutex // for exclusive access
}

func (c *Multiplex) LockChannel(channelId uint) bool {
	// lock the Multiplex, but return 0 only if the given channel exists
	c.Lock()

	if c.channels[channelId] == nil {
		c.Unlock()
		return false
	}

	return true
}

// ----------------------------------------------------------------------
//
//   UTILS
//
// ----------------------------------------------------------------------
// -- MUTEX

func (c *Multiplex) lock_channel(channelId uint) bool {
	// lock the Multiplex, but return true only if the given channel exists
	c.Lock()

	if c.channels[channelId] == nil {
		c.Unlock()
		return false
	}

	return true
}

// ----------------------------------------------------------------------
//
//   BASICS
//
// ----------------------------------------------------------------------
// -- CREATE
func NewMultiplex(conn net.Conn) *Multiplex {
	return NewMultiplexEx(conn, MAX_CHANNELS)
}

func NewMultiplexEx(conn net.Conn, max_channels uint) *Multiplex {
	if max_channels < 0 || max_channels > MAX_CHANNELS {
		return nil
	}

	return &Multiplex{conn: conn, max_channels: max_channels}
}

// -- ACTIVATE CHANNEL
func (c *Multiplex) enable_channel(channelId uint, initialBufferSize int) {
	if c != nil && channelId >= 0 && channelId <= (c.max_channels-1) && c.channels[channelId] == nil {
		if initialBufferSize <= 0 {
			initialBufferSize = INITIAL_BUFFER_SIZE
		}

		buf := &ChannelBuffer{data: make([]byte, initialBufferSize), initial: initialBufferSize}
		c.channels[channelId] = buf
	}
}

func (c *Multiplex) Enable(channelId uint, initialBufferSize int) {
	c.Lock()
	c.enable_channel(channelId, initialBufferSize)
	c.Unlock()
}

func (c *Multiplex) EnableRange(minChannel, maxChannel uint, initialBufferSize int) {
	c.Lock()
	for i := minChannel; i <= maxChannel; i++ {
		c.enable_channel(i, initialBufferSize)
	}
	c.Unlock()
}

func (c *Multiplex) Disable(channelId uint) {
	if c.lock_channel(channelId) {
		c.channels[channelId] = nil
		c.Unlock()
	}
}

// ----------------------------------------------------------------------
//
//   REALLOCATION
//
// ----------------------------------------------------------------------
// We double the buffer size if necessary, and we reduce it by at least
// half if less than 25% is filled.
func (c *Multiplex) reallocate_channel(channelId uint, additionalDataSize int) bool {
	if c == nil || c.channels[channelId] == nil {
		return false
	}

	buf := c.channels[channelId]
	newLen := buf.offset + buf.length + additionalDataSize
	allocateLen := len(buf.data) // cap() ?

	if allocateLen > newLen*4 { // Case 1: buffer is too empty (less than 25%)
		allocateLen = buf.initial
	} else if allocateLen >= newLen { // Case 2: buffer is big enough
		return true
	} else if allocateLen >= (buf.length + additionalDataSize) { // Case 3: move data within buffer (set offset to 0)
		if buf.offset > 0 {
			copy(buf.data, buf.data[buf.offset:buf.offset+buf.length])
			buf.offset = 0
		}
		return true
	}

	// Case 4: shrink or extend buffer
	for allocateLen < newLen {
		allocateLen *= 2
	}

	newbuf := make([]byte, allocateLen)
	copy(newbuf, buf.data[buf.offset:buf.offset+buf.length])
	buf.data = newbuf
	buf.offset = 0

	return true
}

// ----------------------------------------------------------------------
//
//   MODIFY BUFFER
//
// ----------------------------------------------------------------------
func (c *Multiplex) write_channel(channelId uint, data []byte) {
	length := len(data)

	if c.reallocate_channel(channelId, length) {
		buf := c.channels[channelId]
		if buf != nil {
			copy(buf.data[buf.offset:], data)
			buf.length += length
			buf.newData = length
		}
	}
}

func (c *Multiplex) Write(channelId uint, data []byte) {
	if c.lock_channel(channelId) {
		c.write_channel(channelId, data)
		c.Unlock()
	}
}

func (c *Multiplex) copy_channel(channelId uint, dst []byte) (int, error) {
	buf := c.channels[channelId]
	if buf == nil {
		return 0, CHANNEL_IGNORED
	}

	copyLen := len(dst)

	if buf.length < len(dst) {
		copyLen = buf.length
	}

	copy(dst, buf.data[buf.offset:buf.offset+copyLen])
	return copyLen, nil
}

func (c *Multiplex) Copy(channelId uint, dst []byte) (int, error) {
	if !c.lock_channel(channelId) {
		return 0, CHANNEL_CLOSED
	}

	defer c.Unlock()
	return c.copy_channel(channelId, dst)
}

func (c *Multiplex) read_channel(channelId uint, dst []byte) (int, error) {
	buf := c.channels[channelId]
	if buf == nil {
		return 0, CHANNEL_IGNORED
	}

	copyLen := len(dst)

	if buf.length < copyLen {
		copyLen = buf.length
		dst = dst[:copyLen]
	}

	copy(dst, buf.data[buf.offset:buf.offset+copyLen])
	buf.offset += copyLen
	buf.length -= copyLen
	buf.newData -= copyLen

	if buf.newData < 0 {
		buf.newData = 0
	}
	if buf.length <= 0 {
		buf.length = 0
		buf.newData = 0
		buf.offset = 0
	}

	return copyLen, nil
}

func (c *Multiplex) Read(channelId uint, dst []byte) (int, error) {
	if !c.lock_channel(channelId) {
		return 0, CHANNEL_CLOSED
	}

	defer c.Unlock()
	return c.read_channel(channelId, dst)
}

func (c *Multiplex) clear_channel(channelId uint) {
	buf := c.channels[channelId]
	buf.offset = 0
	buf.length = 0
	buf.newData = 0
}

func (c *Multiplex) Clear(channelId uint) {
	if c.lock_channel(channelId) {
		c.clear_channel(channelId)
		c.Unlock()
	}
}

// ----------------------------------------------------------------------
//
//   RECEIVE LOGIC
//
// ----------------------------------------------------------------------
func conn_read(conn net.Conn, timeout time.Duration, buffer []byte) (int, error) {
	if timeout != time.Duration(0) {
		conn.SetReadDeadline(time.Now().Add(timeout))
	}

	position := 0
	length := len(buffer)

	for position < length {
		bytesRead, err := conn.Read(buffer[position:])
		if err != nil {
			if err == io.EOF {
				log.Println("conn_read", "CLOSED")
				return 0, CHANNEL_CLOSED
			} else if neterr, ok := err.(net.Error); ok && neterr.Timeout() {
				log.Println("conn_read", "TIMEOUT")
				return 0, CHANNEL_TIMEOUT
			} else {
				// we should do better than this
				log.Println("conn_read", err)
				return 0, CHANNEL_CLOSED
			}
		} else {
			position += bytesRead
		}
	}

	return position, nil
}

func (c *Multiplex) select_channel(timeout time.Duration, channelId uint) (uint, error) {
	if c == nil {
		return 0, CHANNEL_CLOSED
	}

	// Check if data is available somewhere
	if channelId < c.max_channels {
		if buf := c.channels[channelId]; buf != nil && buf.length > 0 && buf.newData != 0 {
			buf.newData = 0
			return channelId, nil
		}
	}

	for i := 0; i < int(c.max_channels); i++ {
		if buf := c.channels[i]; buf != nil && buf.length > 0 && buf.newData != 0 {
			buf.newData = 0
			return uint(i), nil
		}
	}

	//
	var prefixBuffer [headerLength]byte
	n, err := conn_read(c.conn, timeout, prefixBuffer[:])
	if err != nil {
		return 0, err
	}
	if n != headerLength {
		log.Println("expected", headerLength, "read", n)
		return 0, CHANNEL_IGNORED
	}

	/*
			if prefixBuffer[0] != magic {
				log.Println("expected", magic, "got", prefixBuffer)
		                return 0, CHANNEL_IGNORED
			}
	*/

	//
	dataLength := int(prefixBuffer[0])<<24 | int(prefixBuffer[1])<<16 | int(prefixBuffer[2])<<8 | int(prefixBuffer[3])<<0
	channelId = uint(prefixBuffer[4])

	buffer := make([]byte, dataLength-1)
	start := 0
	for start < dataLength-1 {
		n, err = conn_read(c.conn, time.Duration(0), buffer[start:])
		if err != nil {
			return 0, err
		}
		if n == 0 {
			log.Println("select_channel", "expected", len(buffer)-start, "got 0")
		}
		start += n
	}

	if c.channels[channelId] == nil {
		return channelId, CHANNEL_IGNORED
	}
	c.write_channel(channelId, buffer)
	return channelId, nil
}

func (c *Multiplex) Select(timeout time.Duration) (uint, error) {
	c.Lock()
	defer c.Unlock()

	return c.select_channel(timeout, c.max_channels)
}

func (c *Multiplex) Ignore(channelId uint) {
	c.Lock()
	c.channels[channelId].newData = 0
	c.Unlock()
}

func (c *Multiplex) receive_channel(timeout time.Duration, channelId uint, dst []byte) (int, error) {
	if c == nil {
		return 0, CHANNEL_CLOSED
	}

	buf := c.channels[channelId]
	if buf == nil {
		return 0, CHANNEL_IGNORED
	}

	length := len(dst)

	// Check if data is already buffered.
	if buf.length > 0 {
		if length >= buf.length {
			length = buf.length
		}

		copy(dst, buf.data[buf.offset:buf.offset+length])
		buf.offset += length
		buf.length -= length
		return length, nil
	}

        receiveId, err := c.select_channel(timeout, channelId)
        if err != nil {
                return 0, err
        }

        if receiveId != channelId {
                return 0, CHANNEL_IGNORED
        }

	// Copy from ChannelBuffer
	return c.read_channel(channelId, dst)
}

func (c *Multiplex) Receive(timeout time.Duration, channelId uint, data []byte) (int, error) {
        for {
	    c.Lock()
	    n, err := c.receive_channel(timeout, channelId, data)
            if err != CHANNEL_IGNORED {
                c.Unlock()
                return n, err
            }

            c.Unlock()
        }

        // unreachable
        return 0, CHANNEL_IGNORED
}

// ----------------------------------------------------------------------
//
//   SEND LOGIC
//
// ----------------------------------------------------------------------
func (c *Multiplex) Send(channelId uint, src []byte) (int, error) {
	if len(src) == 0 {
		return 0, nil
	}

	c.Lock()
	defer c.Unlock()

	length := len(src) + 1

	buffer := []byte{
		//magic,
		(byte)((length >> 24) & 0xFF),
		(byte)((length >> 16) & 0xFF),
		(byte)((length >> 8) & 0xFF),
		(byte)((length >> 0) & 0xFF),
		(byte)(channelId & 0xFF)}

	buffer = append(buffer, src...)

	n, err := c.conn.Write(buffer)
	if n != len(buffer) || err != nil {
		log.Println("sent ", n, "expected", len(buffer), err)
	} else {
		n -= headerLength
	}

	return n, err
}

// ----------------------------------------------------------------------
//
//   BUFFER INSPECTION
//
// ----------------------------------------------------------------------
func (c *Multiplex) Length(channelId uint) int {
	if !c.lock_channel(channelId) {
		return -1
	}

	defer c.Unlock()
	return c.channels[channelId].length
}

func (c *Multiplex) LastReceived(channelId uint) int {
	if !c.lock_channel(channelId) {
		return -1
	}

	defer c.Unlock()
	return c.channels[channelId].newData
}

func (c *Multiplex) Get(channelId uint) []byte {
	c.Lock()
	defer c.Unlock()

	buf := c.channels[channelId]
	return buf.data[buf.offset:]
}

func (c *Multiplex) Dup(channelId uint) []byte {
	c.Lock()
	defer c.Unlock()

	buf := c.channels[channelId]
	return append([]byte(nil), buf.data[buf.offset:buf.offset+buf.length]...)
}
